util: Add Python implementation of terminal client (gem5term)

powerjg opened the pull request:
Add a new Python implementation of the terminal client that connects to gem5's terminal server. This implementation maintains compatibility with the existing C version while adding new features:

- Session logging with timestamps
- Configurable buffer sizes and timeouts
- Terminal window size handling
- Better error messages and connection handling
- Unix domain socket support maintained from C version
- Command line argument parsing with help text

The Python version has been tested with both TCP/IP and Unix domain socket connections. Some behavioral improvements over the C version include:
- More graceful connection error handling
- Better cleanup on exit
- Proper terminal restoration after unexpected disconnects

Known limitations:
- Line wrapping behavior is still imperfect, especially when typing long commands (though historical command display works correctly)
- Terminal size updates only happen at connection time

This implementation serves as a more maintainable alternative to the C version while maintaining backward compatibility. Added documentation in README.md explains the features and usage of both implementations.

Testing: Manual testing with gem5 full system simulations including boot, interactive shell usage, and various terminal conditions.

powerjg committed:
util: Add Python implementation of terminal client (gem5term)

Add a new Python implementation of the terminal client that connects to gem5's
terminal server. This implementation maintains compatibility with the existing
C version while adding new features:

- Session logging with timestamps
- Configurable buffer sizes and timeouts
- Terminal window size handling
- Better error messages and connection handling
- Unix domain socket support maintained from C version
- Command line argument parsing with help text

The Python version has been tested with both TCP/IP and Unix domain socket
connections. Some behavioral improvements over the C version include:
- More graceful connection error handling
- Better cleanup on exit
- Proper terminal restoration after unexpected disconnects

Known limitations:
- Line wrapping behavior is still imperfect, especially when typing long
  commands (though historical command display works correctly)
- Terminal size updates only happen at connection time

This implementation serves as a more maintainable alternative to the C
version while maintaining backward compatibility. Added documentation
in README.md explains the features and usage of both implementations.

Testing: Manual testing with gem5 full system simulations including
boot, interactive shell usage, and various terminal conditions.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg commented:
@mahyarsamani this PR was created based on some of your feedback. Do you have time to review it?

mahyarsamani commented:
Will do. Thanks for the prompt action.

mahyarsamani commented on util/term/README.md:
```diff
@@ -0,0 +1,69 @@
+./term localhost 3456
+
+# Using Python client
+./term.py localhost 3456
```
should this be gem5term?

mahyarsamani commented on util/term/README.md:
```diff
@@ -0,0 +1,69 @@
+# Using Python client
+./term.py localhost 3456
+# or simply (localhost is default)
+./term.py 3456
```
ditto

erin-le committed:
util: fix characters inputted to terminal being printed twice

Previously, there was a visual glitch where characters typed into
the terminal (e.g. as part of a command) would be displayed twice.
This commit fixes this issue.

erin-le committed:
util: update README.md for gem5term

This commit changes gem5term README so ./gem5term is used instead
of ./term for the Python client. It also clarifies how to set up
the unix socket in the gem5 configuration script/

erin-le reviewed (commented):
I think this should be good to go, but since I added some changes, it would be nice if someone else could take a look and test it.

erin-le commented:
Some observations about text wrapping when entering text into the terminal, for future reference: After a certain number of characters, the cursor will jump to the start of the current line and start writing new characters over the ones already there. I tried it with a small window size and saw that it only seemed to happen once, with lines after the overwritten line wrapping correctly. The terminal still picks up all characters that are inputted. 

I also noticed that pressing the up and down arrow keys to access previous commands causes the cursor to move up a line and overlap with previously displayed text.
